File: objects/src/object/vector.rs
use crate::Object;
use type_macros::*;
use types::{
    include_template,
    prelude::*,
    source::{include, type_alias, SourceBuilder, SourceTree},
    Config,
};

#[derive(Clone, Copy, Debug, EntityId, Entity, SizedEntity, EntitySource)]
pub struct Cache<T: SizedEntity> {
    pub index: usize,
    pub inner: T,
}

impl<G: Geometry, T: Object<G>> Object<G> for Vec<T>
where
    Self: Entity,
{
    type Cache = Cache<T::Cache>;

    fn object_name() -> (String, String) {
        (
            format!("Object{}", Self::name().0),
            format!("object_{}", Self::name().1),
        )
    }
    fn object_source(cfg: &Config) -> SourceTree {
        SourceBuilder::new(format!("generated/{}.hh", Self::object_name().1))
            .tree(Self::source(cfg))
            .tree(G::geometry_source(cfg))
            .tree(T::object_source(cfg))
            .tree(Self::Cache::source(cfg))
            .content(&include(&format!(
                "render/light/{}.hh",
                G::geometry_name().1,
            )))
            .content(&type_alias(&Self::object_name().0, &Self::name().0))
            .content(&include_template!(
                "object/vector.inl",
                ("Self", "self") => Self::object_name(),
                ("Base", "base") => Self::name(),
                ("Cache", "cache") => Self::Cache::name(),
                ("Geo", "geo") => G::geometry_name(),
                ("Object", "object") => T::object_name(),
                ("ObjectCache", "object_cache") => T::Cache::name(),
            ))
            .build()
    }
}

File: kernel/ocl/object/test/vector.cpp
#include <types.hh>
#include <geometry/euclidean.hh>
#include <geometry/ray_eu.hh>
#include <material/basic.hh>
#include <render/light/eu.hh>

#include "covered/eu_def.hpp"

#define $Self Vector_CoveredTestEu
#define $self vector__covered_test_eu
#define $Element CoveredTestEu
#define $element covered_test_eu
#define $element_size 0
#include <container/vector.inl>
#undef $Self
#undef $self
#undef $Element
#undef $element
#undef $element_size

typedef struct ObjectVectorCache {
    usize index;
    CoveredTestEuCache inner;
} ObjectVectorCache;

typedef Vector_CoveredTestEu ObjectVector_CoveredTestEu;

#define $Self ObjectVector_CoveredTestEu
#define $self object_vector__covered_test_eu
#define $Base Vector_CoveredTestEu
#define $base vector__covered_test_eu
#define $Cache ObjectVectorCache
#define $cache object_vector_cache
#define $Geo Eu
#define $geo eu
#define $Object CoveredTestEu
#define $object covered_test_eu
#define $ObjectCache CoveredTestEuCache
#define $object_cache covered_test_eu_cache
#include <object/vector.inl>
#undef $Self
#undef $self
#undef $Base
#undef $base
#undef $Cache
#undef $cache
#undef $Geo
#undef $geo
#undef $Object
#undef $object
#undef $ObjectCache
#undef $object_cache
